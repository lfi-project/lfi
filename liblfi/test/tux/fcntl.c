#include <unistd.h>
#include <stdio.h>
#include <fcntl.h>
#include <string.h>

int main() {
    // Test fcntl() syscall
    int fd1 = open("/dev/null", O_WRONLY);
    if (fd1 < 0) {
        printf("failed to open /dev/null\n");
        return 1;
    }
    
    // Test F_DUPFD - duplicate fd with minimum fd number
    int fd2 = fcntl(fd1, F_DUPFD, 5);
    if (fd2 < 0) {
        printf("fcntl F_DUPFD failed\n");
        return 1;
    }
    if (fd2 < 5) {
        printf("fcntl F_DUPFD returned fd %d, expected >= 5\n", fd2);
        return 1;
    }
    
    printf("fcntl F_DUPFD: original fd=%d, dup fd=%d\n", fd1, fd2);
    
    // Test F_GETFL - get file status flags
    int flags = fcntl(fd1, F_GETFL);
    if (flags < 0) {
        printf("fcntl F_GETFL failed\n");
        return 1;
    }
    
    printf("fcntl F_GETFL: fd=%d has flags=0x%x\n", fd1, flags);
    
    // Test F_GETFD - get file descriptor flags
    int fd_flags = fcntl(fd1, F_GETFD);
    if (fd_flags < 0) {
        printf("fcntl F_GETFD failed\n");
        return 1;
    }
    
    printf("fcntl F_GETFD: fd=%d has fd_flags=0x%x\n", fd1, fd_flags);
    
    // Test F_SETFD - set file descriptor flags (should work)
    int result = fcntl(fd1, F_SETFD, FD_CLOEXEC);
    if (result < 0) {
        printf("fcntl F_SETFD failed\n");
        return 1;
    }
    
    // Verify the flag was set
    fd_flags = fcntl(fd1, F_GETFD);
    if (fd_flags < 0) {
        printf("fcntl F_GETFD after F_SETFD failed\n");
        return 1;
    }
    
    printf("fcntl F_SETFD: fd=%d now has fd_flags=0x%x\n", fd1, fd_flags);
    
    // Test F_DUPFD_CLOEXEC - should return ENOSYS
    int fd3 = fcntl(fd1, F_DUPFD_CLOEXEC, 8);
    if (fd3 >= 0) {
        printf("fcntl F_DUPFD_CLOEXEC should have failed but returned fd=%d\n", fd3);
        return 1;
    }
    
    printf("fcntl F_DUPFD_CLOEXEC correctly returned error (not supported)\n");
    
    // Test writing to duplicated fd to verify it works
    const char* msg = "test\n";
    write(fd1, msg, strlen(msg));
    write(fd2, msg, strlen(msg));
    
    close(fd1);
    close(fd2);
    
    printf("fcntl test passed\n");
    return 0;
}