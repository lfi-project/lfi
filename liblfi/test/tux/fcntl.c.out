fcntl F_DUPFD: original fd=3, dup fd=5
fcntl F_GETFL: fd=3 has flags=0x20001
fcntl F_GETFD: fd=3 has fd_flags=0x0
fcntl F_SETFD: fd=3 now has fd_flags=0x1
fcntl F_DUPFD_CLOEXEC correctly returned error (not supported)
fcntl test passed
